src/Core/Form/IdentifiableObject/CommandBuilder/CommandBuilder.php

```diff
@@ -60,7 +68,18 @@
  *     $config = new CommandBuilderConfig();
  *     $config
  *         ->addField('[number]', 'setNumber', CommandField::TYPE_INT)
- *         ->addField('[author][name]', 'setNumber', CommandField::TYPE_STRING)
+ *         ->addField('[author][name]', 'setAuthorName', CommandField::TYPE_STRING)
+ *         ->addField(
+ *             '[isValid]',
+ *             'setIsValid',
+ *             CommandField::TYPE_BOOL,
```

Comment:
It's a bit weird that there is always an initial value to rely on (here a boolean), then why is this field more important than the second one? It could have been defined the other way around

What if the setter actually needs three arguments, which one of the three would we use?

I think your initial proposal was cleaner, to add a dedicated method `addFieldWithCallback` for complex use cases like this one, but the method should allow to define more things like:
- what is the property path of other fields?
- what is the type of each field?
- are they all multishop titght together? if so which one needs to have the matching "modify all" checkbox? one of them, all of them, either one of them?

Reply:
I also thought we could add a new command field type for these use cases like `CommandField::TYPE_CALLABLE` or `CommandField::TYPE_MIXED` or `CommandField::TYPE_MULTIPLE_ARGUMENTS`

Reply:
If so maybe we could have a refacto of `CommandField` to do this

```php
CommandField {
    protected function __constructor(
        string $dataPath,
        string $type,
        bool $multiShopField,
        bool $hasDefaultValue = false,
        $defaultValue = null
    )
}
```

We can remove the `$commandSetter` parameter as it could be used in the config field has a key to associate with the `CommandField` object. Also to enclose the possibilities of this object we could change its constructor to a protected one and replace it with static function used as factory methods like:

```php
// No multishop, no default value
CommandField::simpleField(string $dataPath, string $type);
// No multishop, default value defined (we need a boolean and a value container, so that null can be used as a default value)
CommandField::simpleFieldWithDefault(string $dataPath, string $type, $defaultValue); 
// Has multishop, default value defined
CommandField::multishopFieldWithDefault(string $dataPath, string $type, $defaultValue);
```

```php
    ->addComposedField([
        CommandField::simpleField('[isValid]', CommandField::TYPE_BOOL),
        CommandField::simpleFieldWithDefault('[foo]', CommandField::TYPE_STRING, 'default')
    ], `setIsValid`)
```

Reply:
@jolelievre You arte proposing good ideas.
The current implementation was supposed to be the most compatible with current builder strategy. Nothing to change, only an extra callback to handle tricky cases:
- extra arguments passed to the setter are supposed to be form values only ? no argument from outer space ?
- what if the "trigger" field is only a checkbox we don't care about ? and the real stuff is handled by other fields around ?
- the default value is rarely used, because a missing form field is acceptable as partial submission

In my humble opinion, there are too many tricky cases we can't anticipate. Very difficult to propose a consistent set of methods.
I didn't want to over-engineer, but I was initially tempted to replace `CommandBuilderConfig` methods with a strategy pattern with `CommandFieldInterface`:
```php
interface CommandFieldInterface
{
    public function isMutliShop(): bool;

    /**
     * @param array|object $data
     */
    public function exists($data): bool;

    /**
     * @param array|object $data
     */
    public function getValues($data): array;

    /**
     * @param array|object $data
     */
    public function updateCommand(object $command, $data): void;
}

class SomeCommandField implements CommandFieldInterface
{
    protected function __construct(string $setterMethod, string $propertyPath, bool $isMultiShop)
    {
    }

    public static create(string $setterMethod, string $propertyPath): self
    {
        return new static($setterMethod, $propertyPath, false);
    }

    public static createMutliShop(string $setterMethod, string $propertyPath): self
    {
        return new static($setterMethod, $propertyPath, true);
    }

    // Missing interface methods are skipped here
}

class CommandBuilderConfig
{
    public function __construct(CommandFieldInterface ...$fields)
    {
        $this->fields = $fields;
    }
}
```
Now we can use default fields (integer, string, multiple...), or define custom fields:
```php
$config = new CommandBuilderConfig(
    IntegerCommandField::create('setAnInteger', '[path_to_integer]'),
    StringCommandField::createMultiShop('setAString', '[path_to_string]'),
    MultipleCommandField('setMultipleArguments', [
        '[path1]' => 'int',
        '[path2]' => 'string',
    ]),
    MyTrickyCommandField::createMultiShop() // too specific to really care about arguments
);
```
It means we could use default fields in `MyTrickyCommandField::updateCommand`:
1. provide a default value for each field if `exists()` returns `false`
2. invoke their `getValues()` method to collect values otherwise
3. pass the values to the command setter as usual

IMHO switching to strategy pattern is very flexible and extendable, no more closed door.

Reply:
The explicit type casting might not be required if setter arguments are properly type hinted.
Reflection could automatize casting, with lower performance of course... but does it really matter here ?

Reply:
@jolelievre A couple of new methods in `CommandBuilderConfig` could handle cases with multiple fields:
```php
->addMultiShopFields('setRedirectOption', [
    '[seo][redirect_option][type]' => CommandField::TYPE_STRING,
    '[seo][redirect_option][target][id]' => [
        'type' => CommandField::TYPE_INT,
        'default' => 0,
    ],
])
->addFields(/* ... */)
```
In the description array, each field is defined with:
- its path as key
- its type as value, or an array of options: type and default value (optional)

If a single property is missing in the data, Then the mutli-fields is ignored. Defining a default value is required to handle the missing property.

Reply:
Alternative interface with array of options only (no path as key):
```php
->addMultiShopFields('setRedirectOption', [
    [
        'path' => '[seo][redirect_option][type]',
        'type' => CommandField::TYPE_STRING,
    ],
    [
        'path' => '[seo][redirect_option][target][id]',
        'type' => CommandField::TYPE_INT,
        'default' => 0,
    ],
])
```

Reply:
I like the first proposal better personally, and mixing both seems overkill